src/network/network_gui.cpp

```diff
@@ -1564,333 +1575,735 @@ NetworkCompanyInfo *GetLobbyCompanyInfo(CompanyID company)
+						_admin_client_id = this->dd_client_id;
+						text = STR_NETWORK_CLIENT_LIST_ASK_CLIENT_KICK;
+						callback = AdminClientKickCallback;
+						SetDParamStr(0, NetworkClientInfo::GetByClientID(_admin_client_id)->client_name);
+						break;
+
+					case DD_CLIENT_ADMIN_BAN:
+						_admin_client_id = this->dd_client_id;
+						text = STR_NETWORK_CLIENT_LIST_ASK_CLIENT_BAN;
+						callback = AdminClientBanCallback;
+						SetDParamStr(0, NetworkClientInfo::GetByClientID(_admin_client_id)->client_name);
+						break;
+
+					case DD_COMPANY_ADMIN_RESET:
+						_admin_company_id = this->dd_company_id;
+						text = STR_NETWORK_CLIENT_LIST_ASK_COMPANY_RESET;
+						callback = AdminCompanyResetCallback;
+						SetDParam(0, _admin_company_id);
+						SetDParam(1, _admin_company_id);
+						break;
+
+					case DD_COMPANY_ADMIN_UNLOCK:
+						_admin_company_id = this->dd_company_id;
+						text = STR_NETWORK_CLIENT_LIST_ASK_COMPANY_UNLOCK;
+						callback = AdminCompanyUnlockCallback;
+						SetDParam(0, _admin_company_id);
+						SetDParam(1, _admin_company_id);
+						break;
+
+					default:
+						NOT_REACHED();
+				}
 
-		num *= this->line_height;
+				assert(text != STR_NULL);
+				assert(callback != nullptr);
 
-		int diff = (num + WD_FRAMERECT_TOP + WD_FRAMERECT_BOTTOM) - (this->GetWidget<NWidgetBase>(WID_CL_PANEL)->current_y);
-		/* If height is changed */
-		if (diff != 0) {
-			ResizeWindow(this, 0, diff, false);
-			return false;
+				/* Always ask confirmation for all admin actions. */
+				ShowQuery(
+					STR_NETWORK_CLIENT_LIST_ASK_CAPTION,
+					text,
+					this,
+					callback
+				);
+			} break;
+
+			default:
+				NOT_REACHED();
 		}
-		return true;
+
+		this->SetDirty();
 	}
 
-	void UpdateWidgetSize(int widget, Dimension *size, const Dimension &padding, Dimension *fill, Dimension *resize) override
+	void OnQueryTextFinished(char *str) override
 	{
-		if (widget != WID_CL_PANEL) return;
+		if (str == nullptr) return;
 
-		this->server_client_width = std::max(GetStringBoundingBox(STR_NETWORK_SERVER).width, GetStringBoundingBox(STR_NETWORK_CLIENT).width) + WD_FRAMERECT_RIGHT;
-		this->icon_size = GetSpriteSize(SPR_COMPANY_ICON);
-		this->line_height = std::max(this->icon_size.height + 2U, (uint)FONT_HEIGHT_NORMAL);
+		switch (this->query_widget) {
+			default: NOT_REACHED();
 
-		uint width = 100; // Default width
-		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
-			width = std::max(width, GetStringBoundingBox(ci->client_name).width);
-		}
+			case WID_CL_SERVER_NAME_EDIT: {
+				if (!_network_server) break;
+
+				uint index;
+				GetSettingFromName("network.server_name", &index);
+				SetSettingValue(index, StrEmpty(str) ? "My Server" : str);
+				this->InvalidateData();
+				break;
+			}
 
-		size->width = WD_FRAMERECT_LEFT + this->server_client_width + this->icon_size.width + WD_FRAMERECT_LEFT + width + WD_FRAMERECT_RIGHT;
+			case WID_CL_CLIENT_NAME_EDIT: {
+				uint index;
```

Comment:
Please add a `if (!NetworkValidateClientName(str)) break;` here, so we bail out early in case of an invalid name.